File: credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long number = get_long("Number:\n");
    int length=0;
    string output = "INVALID\n";
    bool valid = false;
    long second_to_last = number/10;
    int sum = 0;
    while (second_to_last > 0) {
        int digit = second_to_last % 10;
        if(digit*2 >= 10) {
            sum += (digit*2 % 10) + 1;
        } else {
            sum += digit*2;
        }
        second_to_last /= 100;
    }
    long num = number;
    while (num > 0) {
        int digit = num % 10;
        sum += digit;
        num /= 100;
    }
    if(sum % 10 == 0) {
        valid = true;
    }
    //calc first numbers
    int amex = (number / 10000000000000);
    int mastercard = (number / 100000000000000);
    int visa_short = (number / 1000000000000);
    int visa_long = (number / 1000000000000000);
    if (valid && (amex == 34 || amex == 37)) {
        output = "AMEX\n";
    } else if (valid && (mastercard == 51 || mastercard == 52 || mastercard == 53 || mastercard == 54 || mastercard == 55)) {
        output = "MASTERCARD\n";
    }else if (valid && (visa_short == 4 || visa_long == 4)) {
        output = "VISA\n";
    }
    printf("%s", output);
}

File: credit.py
from cs50 import get_string
import re

def main():
    numb = get_user_input()
    check_card(numb)

#ask user for credit card number
def get_user_input():
    numb = get_string("Number: ")
    return numb

def check_card(numb):
    #return true or false
    credit_card = luhn(numb)
    #check for amex
    if re.search("^3[47]\d{13}$", numb) and credit_card:
        print("AMEX")
    #check for mastercard
    elif re.search("^5[1-5]\d{14}$", numb) and credit_card:
        print("MASTERCARD")
    #check for visa
    elif (re.search("^4\d{12}$", numb) or re.search("^4\d{15}$", numb)) and credit_card:
        print("VISA")
    #output for invalid credit card number
    else:
        print("INVALID")

#check if it is a credit card
def luhn(input):
    i = len(input)-2
    j = len(input)-1
    sum = 0
    while i >= 0:
        t = int(input[i])*2
        while t:
            sum += t % 10
            t //= 10
        i -= 2
    while j >= 0:
        sum += int(input[j])
        j -= 2
    print(f"SUM: {sum}")
    if sum % 10 == 0:
        return True
    return False

#call main function
main()